<h1 align="center" style="font-size:200%; color:blue">Microsoft’s entrance to Movie business</h1><br>
<h1 align="center" style="font-size:200%; color:blue">Data loading and cleaning</h1> 
<p style="font-size:200%; color:green">Methodology:</p>
<p style="color:black"> ROSEMED method </p> 

# Import libraries

In [1]:
#importing libraries
import pandas as pd 
import numpy as np

## data loading and cleaning

In [2]:
region=pd.read_csv('Data/region.csv')
revenue=pd.read_csv('Data/revenue.csv')
genres=pd.read_csv('Data/genres.csv')
date=pd.read_csv('Data/date.csv')
# credit: Jesse Newman for web scraping

In [3]:
#display(region.head(2))
#display(revenue.head(2))
#display(genres.head(2))
#display(date.head(2))

In [4]:
display(region.loc[region.imdb_id.isnull()==True])
display(revenue.imdb_id.isnull().sum())
display(genres.imdb_id.isnull().sum())
display(date.imdb_id.isnull().sum())

imdb_id title year director production_co region_code rank budget_usd  \
7850     NaN   NaN  NaN      NaN           NaN         NaN  NaN        NaN   

     us_gross  
7850      NaN

0

0

0

In [5]:
region = region.drop(7850)

In [6]:
# removing redundant cols 
region.drop(['director', 'production_co', 'rank', 'budget_usd', 'us_gross'], axis = 1, inplace=True)
#region = region.imdb_id.duplicated(keep='first')
display(region.head(2))

revenue.drop('rank', axis = 1, inplace=True)

imdb_id                                       title  year region_code
0  tt2488496  Star Wars: Episode VII - The Force Awakens  2015        [US]
1  tt4154796                           Avengers: Endgame  2019        [US]

In [7]:
# clean region_code
region['region_code'] = region['region_code'].str.replace('\[', '')
region['region_code'] = region['region_code'].str.replace('\]', '')
display(region.head(2))

imdb_id                                       title  year region_code
0  tt2488496  Star Wars: Episode VII - The Force Awakens  2015          US
1  tt4154796                           Avengers: Endgame  2019          US

In [8]:
region.loc[region.region_code.isnull() == True]

imdb_id                                      title  year region_code
320    tt0142342                                  Big Daddy  1999         NaN
432    tt0119217                          Good Will Hunting  1997         NaN
441    tt0120685                                   Godzilla  1998         NaN
446    tt0100758               Teenage Mutant Ninja Turtles  1990         NaN
491    tt0190332             Crouching Tiger, Hidden Dragon  2000         NaN
...          ...                                        ...   ...         ...
14677  tt0094671                       Apprentice to Murder  1988         NaN
14680  tt0099053                                  Archangel  1990         NaN
14681  tt0129758                          Arguing the World  1997         NaN
14682  tt6097926  Arrangiarsi: Pizza... & the Art of Living  2017         NaN
14699  tt0109191                                  Babyfever  1994         NaN

[1253 rows x 4 columns]

In [9]:
region.region_code.fillna(value = 'unknown', inplace = True) # some of the movies are obviously made in US. Thus I labelled it 
# as 'unknown' so that they maybe included in the analysis once other method can be used to verify their origin

In [10]:
# clean revenue year column 
cnt = 0
for index in range(len(revenue)):
    try:
        clean_year=revenue.iloc[index]['year']
        int(clean_year)
    except ValueError:
        #print(index, revenue.iloc[index]['year'])
        # clean it.
        #clean_year=clean_year.split("-",1)
        revenue.at[index, 'year'] = str(clean_year[0:4])
        #revenue.iloc[index]['year'] = str(clean_year[0:4])
        #print("new year =",clean_year[0:4])
    cnt+=1

revenue['year'] = revenue['year'].astype('int64')
revenue['us_gross'] = revenue['us_gross'].astype('float64')

In [11]:
genres[genres.genres == np.nan]

Empty DataFrame
Columns: [imdb_id, title, year, genres]
Index: []

In [12]:
genres[genres.genres.isnull() == True]

Empty DataFrame
Columns: [imdb_id, title, year, genres]
Index: []

In [13]:
#display(region.info())
#display(revenue.info())
#display(genres.info())
#display(date.info())

In [14]:
#print(region.imdb_id.nunique())
#print(revenue.imdb_id.nunique())
#print(genres.imdb_id.nunique())
#print(date.imdb_id.nunique())

## is region important ? 
Since Microsoft is a US based company it may be predent to select US based movies. Let's inspect the data more closely. 

In [15]:
region.loc[region.region_code != 'US'].head(10)

imdb_id                                        title  year region_code
35   tt2283362               Jumanji: Welcome to the Jungle  2017          PH
109  tt3470600                                         Sing  2016          JP
149  tt3183660      Fantastic Beasts and Where to Find Them  2016          GB
181  tt1727824                            Bohemian Rhapsody  2018          GB
216  tt2379713                                      Spectre  2015          GB
232  tt5013056                                      Dunkirk  2017          GB
280  tt0090555                             Crocodile Dundee  1986          AU
320  tt0142342                                    Big Daddy  1999     unknown
331  tt0246460                              Die Another Day  2002          GB
335  tt4123430  Fantastic Beasts: The Crimes of Grindelwald  2018          GB

The 1^st random example <b>"Jumanji: Welcome to the Jungle"</b>  designated as PH in this data is clearly wrong. It's a US based movie and became quite popular critically as well as commercially. Also <b>"spectre"</b>, the 007 series moview is a hugely popular movie. It seems from the first glasnce it may be mileading to filter out movies that are not designated as US.

## Movie budget and gross revenue 

In [16]:
# budget_usd   us_gross
#display(revenue.loc[revenue.budget_usd == -1]['budget_usd'].count())
display(revenue.loc[revenue.budget_usd == -1])
display(revenue.budget_usd.describe())
display(revenue.loc[revenue.us_gross < 0]['us_gross'].count())
display(revenue.us_gross.describe())

imdb_id                      title  year               director  \
304    tt0486946                  Wild Hogs  2007            Walt Becker   
559    tt0082846             On Golden Pond  1981            Mark Rydell   
595    tt0091326     The Karate Kid Part II  1986       John G. Avildsen   
712    tt0034492                      Bambi  1942            James Algar   
726    tt0081562                 Stir Crazy  1980         Sidney Poitier   
...          ...                        ...   ...                    ...   
14421  tt7615722  The Legend of Hallowaiian  2018  Sean Patrick O'Reilly   
14423  tt1757742              Apartment 143  2011         Carles Torrens   
14425  tt0462302               Intervention  2007         Mary McGuckian   
14427  tt4796122                    Satanic  2016        Jeffrey G. Hunt   
14429  tt1157631                Dog Eat Dog  2008          Carlos Moreno   

                       production_co  budget_usd     us_gross  
304              Touchstone Pictures        -1.0  168000000.0  
559                        IPC Films        -1.0  119000000.0  
595                Columbia Pictures        -1.0  115000000.0  
712    Walt Disney Animation Studios        -1.0  102000000.0  
726                Columbia Pictures        -1.0  101000000.0  
...                              ...         ...          ...  
14421                  Arcana Studio        -1.0        451.0  
14423              Nostromo Pictures        -1.0        383.0  
14425                    Scion Films        -1.0        279.0  
14427               Magnet Releasing        -1.0        252.0  
14429                     64 A Films        -1.0         80.0  

[6534 rows x 7 columns]

count    1.443100e+04
mean     4.908767e+08
std      1.154901e+10
min     -1.000000e+00
25%     -1.000000e+00
50%      1.000000e+06
75%      1.800000e+07
max      8.198500e+11
Name: budget_usd, dtype: float64

0

count    1.443100e+04
mean     2.052120e+07
std      4.948649e+07
min      3.000000e+01
25%      1.010000e+05
50%      1.500000e+06
75%      1.900000e+07
max      9.370000e+08
Name: us_gross, dtype: float64

To understand the profitability of the movie both gross revenue and budget is needed. In this dataset with 14,431 entries, 6,534 entries do not have budget information. However, movies like <b>Wild Hogs</b> (\$ 168 million) and <b>Bambi</b> (\$ 102 million) clearly had boxoffice success. Thus, eliminating those data at this stage might not be predent. 

## Movie runtimes 

In [17]:
tbasics = pd.read_csv('Data/imdb.title.basics.csv')  #146144 entries
tbasics.rename(columns = {'tconst':'imdb_id'}, inplace = True)
#print(tbasics.isna().sum())
# runtime_minutes(31739), genres(5408)  missing values but maybe useful still
tbasics['runtime_minutes'].fillna(value=-1.0,inplace=True)
tbasics['runtime_minutes'] = pd.to_numeric(tbasics['runtime_minutes'],errors='coerce')
#tbasics.info()
#display(tbasics.head(2))

## IMDB ratings

In [18]:
ratings = pd.read_csv('Data/imdb.title.ratings.csv')
ratings.rename(columns = {'tconst':'imdb_id'}, inplace = True) 
#ratings.info() # 73856 entries. all data present.
ratings['averagerating'].fillna(value=-1.0,inplace=True)
ratings['averagerating'] = pd.to_numeric(ratings['averagerating'],errors='coerce')
#display(ratings.head(2))

## Join region, genre, date, runtime and rating to  revenue DataFrame

In [19]:
# joining region to the revenue table 
region.drop(['title', 'year'], axis = 1, inplace=True)
region.sort_values(by=['imdb_id'], inplace=False)
revenue.sort_values(by=['imdb_id'], inplace=False)

# join gave a weird bug. joined table didn't have the correct number of records
# so following code joining the tables brute force correctly
j=0
Row_list =[] 
mylist = []
for i in range(len(revenue)):
    #i=j
    while  revenue.iloc[i]['imdb_id'] != region.iloc[j]['imdb_id']:
        j+=1
    if revenue.iloc[i]['imdb_id'] == region.iloc[j]['imdb_id']:
        imdbid=region.iloc[j]['imdb_id']
        code=region.iloc[j]['region_code']
        mylist=[imdbid, code]
        Row_list.append(mylist) 

df = pd.DataFrame(Row_list, columns= region.columns)

# now the join should exactly teild 14,431 records 
df.set_index('imdb_id', inplace=True)
revenue.set_index('imdb_id', inplace=True)
revenue = revenue.join(df, how='right', lsuffix='1', rsuffix='1',  sort=True)
df.reset_index('imdb_id', inplace=True)
revenue.reset_index('imdb_id', inplace=True)

In [20]:
# joining genres to the revenue table 
genres.drop(['title', 'year'], axis = 1, inplace=True)
genres.sort_values(by=['imdb_id'], inplace=False)
revenue.sort_values(by=['imdb_id'], inplace=False)
genres.set_index('imdb_id', inplace=True)
revenue.set_index('imdb_id', inplace=True)
revenue = revenue.join(genres, how='left', lsuffix='1', rsuffix='1',  sort=True)
genres.reset_index('imdb_id', inplace=True)
revenue.reset_index('imdb_id', inplace=True)
#display(revenue.head(3))

In [21]:
# joining date to the revenue table 
date.drop(['title'], axis = 1, inplace=True)
date.sort_values(by=['imdb_id'], inplace=False)
revenue.sort_values(by=['imdb_id'], inplace=False)
date.set_index('imdb_id', inplace=True)
revenue.set_index('imdb_id', inplace=True)
revenue = revenue.join(date, how='left', lsuffix='1', rsuffix='1',  sort=True)
date.reset_index('imdb_id', inplace=True)
revenue.reset_index('imdb_id', inplace=True)
# clean date
revenue['date'] = revenue['date'].astype('datetime64[ns]')
display(revenue.head(3))

imdb_id                                title  year              director  \
0  tt0012190  The Four Horsemen of the Apocalypse  1921            Rex Ingram   
1  tt0012494                              Destiny  1921            Fritz Lang   
2  tt0015648                  Battleship Potemkin  1925  Sergei M. Eisenstein   

                production_co  budget_usd   us_gross region_code  \
0  Metro Pictures Corporation    800000.0  9200000.0          US   
1            Decla-Bioscop AG        -1.0    12000.0          DE   
2                     Goskino        -1.0    51000.0        SUHH   

                   genres       date  
0       Drama,Romance,War 1921-03-06  
1  Drama,Fantasy,Thriller 1921-10-06  
2  Drama,History,Thriller 1925-12-24

In [22]:
# joining runtime to the revenue table 
tbasics.drop(['primary_title', 'original_title', 'start_year', 'genres'], axis = 1, inplace=True)
tbasics.sort_values(by=['imdb_id'], inplace=False)
revenue.sort_values(by=['imdb_id'], inplace=False)
tbasics.set_index('imdb_id', inplace=True)
revenue.set_index('imdb_id', inplace=True)
revenue = revenue.join(tbasics, how='left', lsuffix='1', rsuffix='1',  sort=True)
tbasics.reset_index('imdb_id', inplace=True)
revenue.reset_index('imdb_id', inplace=True)
#display(revenue.head(3))

In [23]:
# joining ratings to the revenue table 
ratings.drop(['numvotes'], axis = 1, inplace=True)
ratings.sort_values(by=['imdb_id'], inplace=False)
revenue.sort_values(by=['imdb_id'], inplace=False)
ratings.set_index('imdb_id', inplace=True)
revenue.set_index('imdb_id', inplace=True)
revenue = revenue.join(ratings, how='left', lsuffix='1', rsuffix='1',  sort=True)
ratings.reset_index('imdb_id', inplace=True)
revenue.reset_index('imdb_id', inplace=True)
#display(revenue.head(3))

In [24]:
# above 3 raws lack both budget and genre information. This dropping
revenue=revenue[revenue.genres != '\\N']
revenue

imdb_id                                title  year  \
0      tt0012190  The Four Horsemen of the Apocalypse  1921   
1      tt0012494                              Destiny  1921   
2      tt0015648                  Battleship Potemkin  1925   
3      tt0017136                           Metropolis  1927   
4      tt0018737                        Pandora's Box  1929   
...          ...                                  ...   ...   
14426  tt9748230      Joni 75: A Birthday Celebration  2019   
14427  tt9799960        311 Enlarged to Show Detail 3  2019   
14428  tt9825006                  Avant qu'on explose  2019   
14429  tt9845398                   End of the Century  2019   
14430  tt9896916                   Pilgrim's Progress  2019   

                   director               production_co  budget_usd  \
0                Rex Ingram  Metro Pictures Corporation    800000.0   
1                Fritz Lang            Decla-Bioscop AG        -1.0   
2      Sergei M. Eisenstein                     Goskino        -1.0   
3                Fritz Lang        Universum Film (UFA)         0.0   
4       Georg Wilhelm Pabst                Nero-Film AG        -1.0   
...                     ...                         ...         ...   
14426         Martyn Atkins                     Unknown        -1.0   
14427           Wayne Price                     Unknown        -1.0   
14428        Rémi St-Michel              Christal Films   5453000.0   
14429          Lucio Castro                     Unknown        -1.0   
14430      Robert Fernandez      Cat In The Mill Studio        -1.0   

        us_gross region_code                      genres       date  \
0      9200000.0          US           Drama,Romance,War 1921-03-06   
1        12000.0          DE      Drama,Fantasy,Thriller 1921-10-06   
2        51000.0        SUHH      Drama,History,Thriller 1925-12-24   
3      1200000.0          DE                Drama,Sci-Fi 1927-01-10   
4        53000.0          DE         Crime,Drama,Romance 1929-01-30   
...          ...         ...                         ...        ...   
14426   752000.0     unknown                       Music 2019-02-07   
14427   239000.0     unknown                       Music        NaT   
14428   120000.0          CA                      Comedy 2019-03-01   
14429    66000.0     unknown                       Drama 2019-08-16   
14430  1300000.0          US  Adventure,Animation,Family 2019-04-18   

       runtime_minutes  averagerating  
0                  NaN            NaN  
1                  NaN            NaN  
2                  NaN            NaN  
3                  NaN            NaN  
4                  NaN            NaN  
...                ...            ...  
14426            128.0            8.4  
14427            133.0            8.9  
14428            108.0            6.9  
14429             84.0            7.2  
14430            108.0            7.9  

[14428 rows x 12 columns]

In [25]:
revenue.to_csv('Data/movie_main.csv', index=False)